### PR TITLE
Add grouped sorted collection

### PR DESCRIPTION
**Pull request by Odrakir**
In an effort to make small PRs this is only a small part needed for #2599 

I add a 2 dimensional collection to store lists with sections. Fully tested.

### CODE REVIEW

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+{
+    public struct SectionedIndex
+    {
+        public int Section;
+        public int Row;
+    }
+
+    public class GroupedOrderedList<TItem>
+    {
+        private List<List<TItem>> sections;
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Couldn't these be immutable? If so, do it. If not, why?

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        public IReadOnlyList<IReadOnlyList<TItem>> Items
+            => sections;
+
+        public GroupedOrderedList(Func<TItem, IComparable> orderingKey, Func<TItem, IComparable> groupingKey, IList<TItem> initialItems = null, bool descending = false)
+        {
+            this.orderingKey = orderingKey;
+            this.groupingKey = groupingKey;
+            this.descending = descending;
+
+            if (initialItems == null) {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: No egyptian, mah dude

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                    .OrderBy(orderingKey)
+                    .GroupBy(groupingKey)
+                    .Select(g => g.ToList())
+                    .ToList();
+            }
+        }
+
+        public TItem ItemAt(int section, int row)
+        {
+            return sections[section][row];
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: I like my expression bodied methods

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        public virtual SectionedIndex? RemoveItem(TItem item)
+        {
+            var sectionIndex = sections
+                .IndexOf(g => groupingKey(g.First()).CompareTo(groupingKey(item)) == 0);
+
+            if (sectionIndex == -1) {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Irrelevant braces

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                return null;
+            }
+
+            var rowIndex = sections[sectionIndex].IndexOf(item);
+            if (rowIndex == -1) {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Irrelevant braces

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                return null;
+            }
+
+            sections[sectionIndex].RemoveAt(rowIndex);
+
+            if (sections[sectionIndex].Count == 0) {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Stop the egyptian

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            sections[section].RemoveAt(row);
+            return item;
+        }
+
+        public virtual SectionedIndex InsertItem(TItem item)
+        {
+            var sectionIndex = sections
+                .IndexOf(g => groupingKey(g.First()).CompareTo(groupingKey(item)) == 0);
+
+            if (sectionIndex == -1) {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: What's this tutankhamun nonsense?

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                var insertionIndex = sections.FindLastIndex(g => groupingKey(g.First()).CompareTo(groupingKey(item)) < 0);
+                List<TItem> list = new List<TItem> { item };
+                if (insertionIndex == -1) {
+                    sections.Insert(0, list);
+                    return new SectionedIndex { Section = 0, Row = 0 };
+                } else {
+                    sections.Insert(insertionIndex + 1, list);
+                    return new SectionedIndex { Section = insertionIndex + 1, Row = 0 };
+                }
+            } else {
```
Comment (willsb) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: You've gone all out

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs (hunk 1 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: It could, but I don't really see the advantage here, as it is used only internally and it's exposing a read only list.

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs (hunk 2 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Ok 😞

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs (hunk 3 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Is that mandatory? I don't like them.

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs (hunk 4 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: You are irrelevant!

On a serious note, I think it's safer to leave them, but ok...

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs (hunk 8 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: The `else` in three lines really hurts my eyes...

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+            if (initialItems == null)
+            {
+                sections = new List<List<TItem>> { };
+            }
+            else
+            {
+                sections = initialItems
+                    .OrderBy(orderingKey)
+                    .GroupBy(groupingKey)
```
Comment (sofokli1) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Doesn't `ordering` before `grouping` assume that the `grouping` will keep the ordering? Shouldn't this happen the other way around to make sure?

Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: I think this could potentially be a problem, you are right. But it's not so simple as to invert the calls. Let me try again.

**Comment by Odrakir** [comment]
I can think on many ways of simplifying the descending/ascending ordering selection in Swift. But this is the best I can do with C#, any ideas are welcome.

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        public int Section;
+        public int Row;
+    }
+
+    public class GroupedOrderedList<TItem>
+    {
+        private List<List<TItem>> sections;
+        private Func<TItem, IComparable> orderingKey;
+        private Func<TItem, IComparable> groupingKey;
+        private bool descending;
```
Comment (sofokli1) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Is this really needed? The `orderingKey` comes from outside so it is already providing the ascending/descending functionality?

Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedList.cs]: Not really, the key says which property to compare, but the `descending` parameter says the order.

File: Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs [hunk 11]
```diff
@@ -0,0 +1,597 @@
...
+
+                List<List<int>> expected = new List<List<int>>
+                {
+                    new List<int> { 1, 2, 3 }
+                };
+                CollectionAssert.AreEqual(intCollection.Items, expected);
+            }
+
+            [Fact, LogIfTooSlow]
+            public void CanRemoteLastItem()
```
Comment (daividssilverio) [inline Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: typo `Remote`

File: Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+
+            public TheRemoveItemAtMethod()
+            {
+                intCollection = new GroupedOrderedList<int>(i => i, i => i, i => i.ToString().Length);
+            }
+
+            [Fact, LogIfTooSlow]
+            public void ThrowsIfEmpty()
+            {
+                Action remoteItemAt = () => intCollection.RemoveItemAt(0, 0);
```
Comment (daividssilverio) [inline Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: remove? typo

File: Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+                remoteItemAt.Should().Throw<ArgumentOutOfRangeException>();
+            }
+
+            [Fact, LogIfTooSlow]
+            public void ThrowsIfIndexOutOfRange()
+            {
+                List<int> list = new List<int> { 40, 70, 8, 3, 1, 2 };
+                intCollection.ReplaceWith(list);
+
+                Action remoteItemAt = () => intCollection.RemoveItemAt(1, 4);
```
Comment (daividssilverio) [inline Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: remove? typo

File: Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+                List<List<int>> expected = new List<List<int>>
+                {
+                    new List<int> { 1, 2, 3 },
+                    new List<int> { 40 }
+                };
+                CollectionAssert.AreEqual(intCollection.Items, expected);
+            }
+
+            [Fact, LogIfTooSlow]
+            public void CanRemoteFirstItem()
```
Comment (daividssilverio) [inline Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: remove? typo

Comment (Odrakir) [reply Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: What's wrong with me?

Comment (daividssilverio) [reply Toggl.Foundation.Tests/MvvmCross/Collections/GroupedOrderedListTests.cs]: lack of coffee?

**Comment by Odrakir** [comment]
I had that option at the beginning, but the constructor is getting big and I think it's not going to get used actually, so I removed it.

We can always bring it back in the future if needed.

File: Toggl.Multivac/Extensions/EnumerableExtensions.cs [hunk 15]
```diff
@@ -19,6 +18,26 @@ public static class EnumerableExtensions
...
 
+        public static int IndexOf<T>(this IEnumerable<T> self, T item, Func<T, IComparable> indexKey)
+        {
+            var itemKey = indexKey(item);
+            return self.IndexOf(i => indexKey(i).CompareTo(itemKey) == 0);
+        }
+
+        public static IEnumerable<T> OrderBy<T>(this IEnumerable<T> self, Func<T, IComparable> orderingKey, bool descending)
+        {
+            if (descending)
```
Comment (kornelijepetak) [inline Toggl.Multivac/Extensions/EnumerableExtensions.cs]: Would be nicer as a conditional. Let alone EBM (but this time I will not make you do one). 😄

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs [hunk 16]
```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Toggl.Multivac.Extensions;
+
+namespace Toggl.Foundation.MvvmCross.Collections
+{
+    public struct SectionedIndex
+    {
+        public int Section;
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: You should not use public fields. Convert these into properties. To make them immutable do a constructor and private setters.

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+                {
+                    sections[sectionIndex].Insert(rowIndex + 1, item);
+                    return new SectionedIndex { Section = sectionIndex, Row = rowIndex + 1 };
+                }
+            }
+        }
+
+        private bool areInOrder(TItem ob1, TItem ob2, Func<TItem, IComparable> key)
+        {
+            if (descending)
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: will look better with `?:`

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public SectionedIndex? RemoveItem(TItem item)
+        {
+            var index = IndexOf(item);
+
+            if (!index.HasValue)
+                return null;
+
+            sections[index.Value.Section].RemoveAt(index.Value.Row);
```
Comment (kornelijepetak) [inline Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: lines 79-82 are repeated conceptually at 90-92... consider turning them into a function

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs (hunk 16 above)
Comment (kornelijepetak) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: Or state if you had a good reason for public fields (hint: there aren't any 😄 )

Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: My only reason was this was a struct and I thought they had some other convention (judging by `Point` in Multivac). Now I see every other struct is as you say, so I guess is just `Point` that's wrong.

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs (hunk 17 above)
Comment (Odrakir) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: 🙅‍♂️

File: Toggl.Multivac/Extensions/EnumerableExtensions.cs (hunk 15 above)
Comment (kornelijepetak) [reply Toggl.Multivac/Extensions/EnumerableExtensions.cs]: I am afraid you are dead wrong. 😛 
But also, your `descending` should be `isDescending`.
And I also didn't say one line. I said conditional.

```C#
return isDescending
    ? self.OrderByDescending(orderingKey)
    : self.OrderBy(orderingKey);
```

File: Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs (hunk 16 above)
Comment (kornelijepetak) [reply Toggl.Foundation.MvvmCross/Collections/GroupedOrderedCollection.cs]: Yes, then `Point` is wrong.

File: Toggl.Multivac/Extensions/EnumerableExtensions.cs (hunk 15 above)
Comment (Odrakir) [reply Toggl.Multivac/Extensions/EnumerableExtensions.cs]: Mppfmmfpd.... ok

Comment (Odrakir) [reply Toggl.Multivac/Extensions/EnumerableExtensions.cs]: *changes code while being grumpy in spanish*
